Replace exceptions related to event subscriptions with either

korthout opened the pull request:
## Description

<!-- Please explain the changes you made here. -->
This replaces the subscribe to message and timer events logic with functional style using the Either type.

To achieve that, this PR adds to the previously added `Either.collect()` a `Either.collectFoldingLeft` to collect multiple calculations (evaluations of the expressions) that can fail to return either a failure or subscribe to all events using the evaluated expressions.

It's likely best to review this on a commit by commit basis. However, please keep in mind that not all changes made in the first commits are kept in the final result.

## Related issues

<!-- Which issues are closed by this PR or are related -->

relates #4323

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [x] I've reviewed my own code
* [x] I've written a clear changelist description
* [x] I've narrowly scoped my changes
* [x] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

korthout commented:
@saig0 Please have a look, and let me know if you see ways to improve this further. In addition, if you want to me extract the Either.peek to a separate pull request, please let me know.

saig0 reviewed (changes requested):
I like the new style :+1: Thanks @korthout :cake: 

I just have some minor comments. Please have a look.

korthout commented on engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java:
```diff
@@ -90,45 +90,54 @@ public void unsubscribeFromEvents(
+        .filter(event -> event.isTimer() || event.isMessage())
+        .map(event -> evalExpressions(event, context))
+        .collect(Either.collector())
+        .peek(results -> subscribeToMessageEvents(context, sideEffects, results))
+        .peek(results -> subscribeToTimerEvents(context, sideEffects, commandWriter, results))
```
That's an interesting point. I guess the confusion during yesterday's show and tell was partially due to the peek and could've been avoided this way. 

I do feel it takes the power of chaining away, which feels very useful in the java world where side-effects are everywhere. Another idea is to change the Either.ifRight signature to what peek actually is. This allows something like the following.

```suggestion
        .ifRight(results -> subscribeToMessageEvents(context, sideEffects, results))
        .ifRight(results -> subscribeToTimerEvents(context, sideEffects, commandWriter, results))
```

 To be fair, probably it's best not allow this chaining altogether and make side effects more visible by having to either place it in the end of a chain, or separate it like in your example. I'll remove the `peek` for now

saig0 replied:
> probably it's best not allow this chaining altogether and make side effects more visible by having to either place it in the end of a chain

I agree. 

Allow to chain `ifRight()` would also differ from the Java API of `Optional#ifPresent()` and the semantics of terminal stream operations.

korthout commented on engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java:
```diff
@@ -229,7 +237,17 @@ public ExpressionProcessor(
     final var expectedTypes = Set.of(ResultType.STRING, ResultType.NUMBER);
     return evaluateExpressionAsEither(expression, scopeKey)
-        .flatMap(result -> typeCheck(result, expectedTypes, scopeKey))
+        .flatMap(
+            result ->
+                typeCheck(result, expectedTypes, scopeKey)
+                    .mapLeft(
+                        failure ->
+                            new Failure(
+                                String.format(
+                                    "Failed to extract the correlation key for '%s': The value must be either a string or a number, but was %s.",
+                                    expression.getExpression(), result.getType()),
+                                ErrorType.EXTRACT_VALUE_ERROR,
+                                scopeKey)))
```
@saig0 Note that this is the only place where we deviate from the standard typecheck failure message. If we align the messages this entire mapLeft is unnecessary. WDYT?

saig0 replied:
Yes, I see your point. I would like to be consistent too.

The advantage of this message is that it says what is wrong. For example, if the incident is created for a task with multiple boundary events. 

We could change the standard type check and add a context to it (e.g. `... while evaluating the correlation key`).

korthout replied:
Interesting. I'll just open an issue for that idea. I'll extract the special code in a separate method for now

korthout commented:
@saig0 I've played around a bit more after your comments. I've added a collectorFoldingLeft Collector in an effort to reduce the required logic, removed the peek method to highlight the side effects and reduced the return value mapping logic to the bare minimum.

Please have another look.

saig0 reviewed (approved):
Looks good :+1:

korthout commented:
bors merge

korthout commented:
bors cancel

korthout commented:
bors merge

korthout commented:
bors retry
